[Merged by Bors] - Use the forwards iterator more often

paulhauner opened the pull request:
## Issue Addressed

NA

## Primary Change

When investigating memory usage, I noticed that retrieving a block from an early slot (e.g., slot 900) would cause a sharp increase in the memory footprint (from 400mb to 800mb+) which seemed to be ever-lasting.

After some investigation, I found that the reverse iteration from the head back to that slot was the likely culprit. To counter this, I've switched the `BeaconChain::block_root_at_slot` to use the forwards iterator, instead of the reverse one.

I also noticed that the networking stack is using `BeaconChain::root_at_slot` to check if a peer is relevant (`check_peer_relevance`). Perhaps the steep, seemingly-random-but-consistent increases in memory usage are caused by the use of this function.

Using the forwards iterator with the HTTP API alleviated the sharp increases in memory usage. It also made the response much faster (before it felt like to took 1-2s, now it feels instant).

## Additional Changes

In the process I also noticed that we have two functions for getting block roots:

- `BeaconChain::block_root_at_slot`: returns `None` for a skip slot.
- `BeaconChain::root_at_slot`: returns the previous root for a skip slot.

I unified these two functions into `block_root_at_slot` and added the `WhenSlotSkipped` enum. Now, the caller must be explicit about the skip-slot behaviour when requesting a root. 

Additionally, I replaced `vec![]` with `Vec::with_capacity` in `store::chunked_vector::range_query`. I stumbled across this whilst debugging and made this modification to see what effect it would have (not much). It seems like a decent change to keep around, but I'm not concerned either way.

Also, `BeaconChain::get_ancestor_block_root` is unused, so I got rid of it :wastebasket:.

## Additional Info

I haven't also done the same for state roots here. Whilst it's possible and a good idea, it's more work since the fwds iterators are presently block-roots-specific.

Whilst there's a few places a reverse iteration of state roots could be triggered (e.g., attestation production, HTTP API), they're no where near as common as the `check_peer_relevance` call. As such, I think we should get this PR merged first, then come back for the state root iters. I made an issue here https://github.com/sigp/lighthouse/issues/2377.

paulhauner committed:
Use fwds iter more often

paulhauner commented:
Results on mainnet are looking quite promising so far! The blue line gets this update shortly before 16:00.

![fwd-iter](https://user-images.githubusercontent.com/6660660/120084386-196c4f00-c113-11eb-87e5-2f773f1d2723.png)

paulhauner commented:
There's also a nice reflection in the number of bytes read from the database:

![disk](https://user-images.githubusercontent.com/6660660/120087958-ae7d4100-c12f-11eb-929c-9646e1db1ab8.png)

And also the count of read operations (ignore the kB scale, it's just a count of operations):

![disk-count](https://user-images.githubusercontent.com/6660660/120088001-e4222a00-c12f-11eb-80e6-7b74923a9757.png)

paulhauner committed:
Remove unused ancestor block root fn

paulhauner commented:
> Results on mainnet are looking quite promising so far! The blue line gets this update shortly before 16:00.

After almost 2 days this node is still showing the same memory usage!

paulhauner committed:
Tidy, specify skips for block_at_slot

paulhauner committed:
Define capacity for chunks range query

paulhauner committed:
Add test

paulhauner committed:
Rename Skips -> WhenSlotSkipped

paulhauner committed:
Handle a skipped head slot

paulhauner committed:
Fix typo

paulhauner committed:
Add fast-lookups

paulhauner commented:
In my last commit (https://github.com/sigp/lighthouse/pull/2376/commits/6c73e22b6a9429179778890d635ea5a77f09961d) I added a "fast lookup" paths. It makes things a little more complicated, but it should save a bunch of overhead when doing queries near the head (e.g., reversing the roots for the fwds iterator).

I think it's worth it, but open to suggestion :)

paulhauner committed:
Tidy fast path

michaelsproul commented on beacon_node/store/src/chunked_vector.rs:
```diff
@@ -431,9 +431,15 @@ fn range_query<S: KeyValueStore<E>, E: EthSpec, T: Decode + Encode>(
 ) -> Result<Vec<Chunk<T>>, Error> {
-    let mut result = vec![];
+    let range = start_index..=end_index;
+    let len = range
```
I guess you were angling for `ExactSizeIterator::len`? I think given that it doesn't exist for `RangeInclusive<usize>` due to overflow we may as well write it manually like:

```suggestion
    let len = end_index.saturating_add(1).saturating_sub(start_index);
```

paulhauner replied:
> I guess you were angling for ExactSizeIterator::len? I think given that it doesn't exist for RangeInclusive<usize> due to overflow we may as well write it manually like:

I was trying to compute the length based upon the range, mainly as a defensive "one source of truth" strategy (i.e., can't change the range and forget to change the length).

paulhauner replied:
I don't really mind either way here. If you would prefer to define the range later then I'm fine to change it back.

michaelsproul replied:
Nah that's fine, lets leave it

paulhauner committed:
Appease clippy

paulhauner committed:
Rename target_slot -> request_slot

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -521,20 +521,119 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+                        return (curr_root != prev_root).then(|| curr_root);
+                    }
+                }
+                prev_root_opt = Some(curr_root);
```
This is a minor style thing, but I think this loop could be more neatly written with `tuple_windows` like:

```rust
for ((prev_root, _), (curr_root, curr_slot)) in iter.tuple_windows() {
    if curr_slot == request_slot {
        return (curr_root != prev_root).then(|| curr_root);
    }
}
```

Further, seeing as we know that the request should be the first item of the `tuple_windows` iterator, we could use `next()`, with an additional check on the slot:

```rust
if let Some((prev_root, _), (curr_root, curr_slot)) = process_results(self.forwards_iter_block_roots(prev_slot)?, |iter| {
    iter.tuple_windows().next()
}) {
    if curr_slot != request_slot {
        return Err(..);
    }
    return Ok((curr_root != prev_root).then(|| curr_root));
} else {
    return Ok(None);
}
```

Idk though, that's a bit messier.

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -521,20 +521,119 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        }
+
+        process_results(self.forwards_iter_block_roots(request_slot)?, |mut iter| {
+            iter.next().map(|(root, _)| root)
```
I think we should check the slot here, analogous to above.

michaelsproul reviewed (commented):
Great find! It's so satisfying to see that memory usage graph drop, and to know the root cause at last!

I've carefully reviewed the fast path and I'm confident that it's correct. I have a few tips for improving readability further.

---

> It makes things a little more complicated, but it should save a bunch of overhead when doing queries near the head (e.g., reversing the roots for the fwds iterator).

The fast path won't help the forwards iterator, because the forwards iterator _always_ clones the head state, and then uses a regular backwards iterator on it, here:

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/beacon_node/store/src/forwards_iter.rs#L72-L81

That state gets passed in when the "hybrid" forwards iterator gets constructed:

https://github.com/sigp/lighthouse/blob/3a24ca5f14c6e6d6612fd43eca82aa0c1e6aba16/beacon_node/store/src/forwards_iter.rs#L97-L103

Even so, I think the fast path optimisation is worth keeping for other cases where we query close to the head state.

EDIT: realised I misinterpreted what you were saying

paulhauner committed:
Implement slot checks, tuple windows

paulhauner commented:
Thanks for the readability tips and the extra sanity checks for `Slot`. I've addressed your comments but haven't done anything for https://github.com/sigp/lighthouse/pull/2376#discussion_r642175222 yet.

It's good to go assuming you don't have more changes :)

michaelsproul reviewed (approved):
LGTM! :rocket:

paulhauner commented:
bors r+

paulhauner commented:
bors r-

(Batching)

paulhauner commented:
bors r+

paulhauner commented:
bors r-

paulhauner committed:
Appease clippy

paulhauner commented:
bors r+

paulhauner commented:
bors r+

paulhauner commented:
bors -

paulhauner commented:
bors r-

paulhauner commented:
bors r+

:crossed_fingers:
